File: packages/ui/src/services/history.ts
import { safeInvoke, isTauri, isVSCode } from '@utils/tauri'
import { postVSCodeMessage } from './vscode-bridge'
import type { QueryHistoryItem } from '@types'

export const historyService = {
    async getQueryHistory(limit?: number, offset?: number): Promise<QueryHistoryItem[]> {
        if (isVSCode()) return postVSCodeMessage<QueryHistoryItem[]>('get_query_history', { limit, offset })
        if (isTauri()) return safeInvoke('get_query_history', { limit, offset }) as Promise<QueryHistoryItem[]>
        return []
    },

    async searchHistory(query: string, limit?: number): Promise<QueryHistoryItem[]> {
        if (isVSCode()) return postVSCodeMessage<QueryHistoryItem[]>('search_history', { query, limit })
        if (isTauri()) return safeInvoke('search_history', { query, limit }) as Promise<QueryHistoryItem[]>
        return []
    },

    async deleteHistoryItem(id: string): Promise<void> {
        if (isVSCode()) return postVSCodeMessage<void>('delete_history_item', { id })
        if (isTauri()) return safeInvoke('delete_history_item', { id }) as Promise<void>
    },

    async clearHistory(connectionId?: string): Promise<number> {
        if (isVSCode()) return postVSCodeMessage<number>('clear_history', { connectionId })
        if (isTauri()) return safeInvoke('clear_history', { connectionId }) as Promise<number>
        return 0
    },
}

File: apps/desktop/src-tauri/src/commands/history.rs
use std::sync::Arc;

use tauri::State;

use crate::core::history_store::HistoryStore;
use crate::models::history::{QueryHistoryFilter, QueryHistoryItem};
use crate::utils::error::AppResult;

#[tauri::command]
pub async fn get_query_history(
    history_store: State<'_, Arc<HistoryStore>>,
    limit: Option<usize>,
    offset: Option<usize>,
) -> AppResult<Vec<QueryHistoryItem>> {
    let filter = QueryHistoryFilter {
        limit: limit.unwrap_or(100),
        offset: offset.unwrap_or(0),
        ..Default::default()
    };

    history_store.get_history(&filter)
}

#[tauri::command]
pub async fn search_history(
    history_store: State<'_, Arc<HistoryStore>>,
    query: String,
    limit: Option<usize>,
) -> AppResult<Vec<QueryHistoryItem>> {
    history_store.search_history(&query, limit.unwrap_or(50))
}

#[tauri::command]
pub async fn delete_history_item(
    history_store: State<'_, Arc<HistoryStore>>,
    id: String,
) -> AppResult<()> {
    history_store.delete_history_item(&id)
}

#[tauri::command]
pub async fn clear_history(
    history_store: State<'_, Arc<HistoryStore>>,
    connection_id: Option<String>,
) -> AppResult<usize> {
    history_store.clear_history(connection_id.as_deref())
}
